### PR TITLE
add pre/post hook to Themis validate

### PR DESCRIPTION
**Pull request by drewwells**
Allow external hooks into Validate for things like request metrics.

### CODE REVIEW

**Comment by fblaha-infoblox** [comment]
LGTM but instead of custom hooks implementation can we use GRPC interceptors for the same purpose as hooks are intended?

File: pdpserver/server/server_test.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	if e := newMissingPolicyError().Error(); !strings.Contains(string(outMsg.Body), e) {
+		t.Errorf("got: %s wanted: %s", string(outMsg.Body), e)
+	}
+}
+
+func TestValidate_hook_context(t *testing.T) {
+
+	var (
+		preCalled bool
+		called    bool
```
Comment (rchowinfoblox) [inline pdpserver/server/server_test.go]: why not "postCalled"?

File: pdpserver/server/server_test.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+		hasKey    bool
+		key       int = 1
+	)
+
+	preHook := func(ctx context.Context) context.Context {
+		preCalled = true
+		return context.WithValue(ctx, key, struct{}{})
+	}
+
+	hook := func(ctx context.Context, msg *pb.Msg, err error) {
```
Comment (rchowinfoblox) [inline pdpserver/server/server_test.go]: why not "postHook"?

File: pdpserver/server/server_test.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if e := true; hasKey != e {
+		t.Error("key was not passed from pre hook to post hook")
+	}
+
+	if e := true; preCalled != e {
+		t.Error("hook was not called")
```
Comment (rchowinfoblox) [inline pdpserver/server/server_test.go]: "preHook was not called"

File: pdpserver/server/server_test.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	}
+
+	if e := true; called != e {
+		t.Error("hook was not called")
```
Comment (rchowinfoblox) [inline pdpserver/server/server_test.go]: "postHook was not called"

File: pdpserver/server/service.go [hunk 5]
```diff
@@ -154,24 +154,41 @@ func (s *Server) rawValidateToBuffer(p *pdp.PolicyStorage, c *pdp.LocalContentSt
 
 // Validate is a server handler for gRPC call
 // It handles PDP decision requests
+// Return variables are named, so they can be passed to validate
```
Comment (rchowinfoblox) [inline pdpserver/server/service.go]: good, i think it make code cleaner and easier to understand

**Pull request merged**